hudi-client/src/main/java/org/apache/hudi/table/upgrade/UpgradeDowngrade.java

```diff
@@ -0,0 +1,153 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.table.upgrade;
+
+import org.apache.hudi.common.table.HoodieTableConfig;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.HoodieTableVersion;
+import org.apache.hudi.common.util.FileIOUtils;
+import org.apache.hudi.config.HoodieWriteConfig;
+
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.spark.api.java.JavaSparkContext;
+
+import java.io.IOException;
+import java.util.Date;
+import java.util.Properties;
+
+/**
+ * Helper class to assist in upgrading/downgrading Hoodie when there is a version change.
+ */
+public class UpgradeDowngrade {
+
+  private static final Logger LOG = LogManager.getLogger(UpgradeDowngrade.class);
+  public static final String HOODIE_UPDATED_PROPERTY_FILE = "hoodie.properties.updated";
+
+  private HoodieTableMetaClient metaClient;
+  private HoodieWriteConfig config;
+  private JavaSparkContext jsc;
+  private transient FileSystem fs;
+  private Path updatedPropsFilePath;
+  private Path propsFilePath;
+
+  /**
+   * Perform Upgrade or Downgrade steps if required and updated table version if need be.
+   * <p>
+   * Starting from version 0.6.0, this upgrade/downgrade step will be added in all write paths.
+   *
+   * Essentially, if a dataset was created using any pre 0.6.0(for eg 0.5.3), and Hoodie version was upgraded to 0.6.0,
+   * Hoodie table version gets bumped to 1 and there are some upgrade steps need to be executed before doing any writes.
+   * Similarly, if a dataset was created using Hoodie version 0.6.0 or Hoodie table version 1 and then hoodie was downgraded
+   * to pre 0.6.0 or to Hoodie table version 0, then some downgrade steps need to be executed before proceeding w/ any writes.
+   *
+   * On a high level, these are the steps performed
+   *
+   * Step1 : Understand current hoodie table version and table version from hoodie.properties file
+   * Step2 : Delete any left over .upgraded from previous upgrade/downgrade
+   * Step3 : If version are different, perform upgrade/downgrade.
+   * Step4 : Copy hoodie.properties -> hoodie.properties.upgraded with the version updated
+   * Step6 : Rename hoodie.properties.updated to hoodie.properties
+   * </p>
+   *
+   * @param metaClient instance of {@link HoodieTableMetaClient} to use
+   * @param toVersion version to which upgrade or downgrade has to be done.
+   * @param config instance of {@link HoodieWriteConfig} to use.
+   * @param jsc instance of {@link JavaSparkContext} to use.
+   * @param instantTime current instant time that should not be touched.
+   */
+  public static void run(HoodieTableMetaClient metaClient, HoodieTableVersion toVersion, HoodieWriteConfig config,
```

Comment:
why use same name for this method and for the other method too?

Reply:
why not?

Reply:
you could use. but at first sight, I got confused that we can calling the same method within run().